File: scripts/delete-all-blobs.py
#!/usr/bin/env python3

"""
Python Delete All Blobs Script for Vercel Blob Storage

This script will delete ALL blobs from your Vercel blob store using HTTP requests.
Use with extreme caution as this action is irreversible!

Usage:
    python scripts/delete-all-blobs.py [--dry-run] [--confirm]

Options:
    --dry-run    Show what would be deleted without actually deleting
    --confirm    Skip confirmation prompt (use for automation)

Requirements:
    pip install requests python-dotenv
"""

import os
import sys
import json
import time
import argparse
import requests
from typing import List, Dict, Any
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class PythonBlobCleaner:
    def __init__(self, dry_run: bool = False, skip_confirmation: bool = False):
        self.token = os.getenv('BLOB_READ_WRITE_TOKEN')
        self.dry_run = dry_run
        self.skip_confirmation = skip_confirmation
        self.batch_size = 50
        self.delay_between_batches = 1.0  # seconds
        
        if not self.token:
            print('❌ Error: BLOB_READ_WRITE_TOKEN environment variable is required')
            sys.exit(1)
    
    def confirm_deletion(self) -> bool:
        """Ask user to confirm deletion"""
        if self.skip_confirmation:
            return True
        
        print('⚠️  This will DELETE ALL BLOBS from your Vercel blob store. This action is IRREVERSIBLE!')
        response = input('Are you absolutely sure you want to continue? Type "DELETE ALL" to confirm: ')
        return response == 'DELETE ALL'
    
    def list_all_blobs(self) -> List[Dict[str, Any]]:
        """List all blobs in the storage"""
        print('📋 Fetching list of all blobs...')
        
        try:
            headers = {
                'Authorization': f'Bearer {self.token}',
                'Content-Type': 'application/json'
            }
            
            response = requests.get(
                'https://api.vercel.com/v2/blob',
                headers=headers,
                params={'limit': '1000'}
            )
            response.raise_for_status()
            
            data = response.json()
            blobs = data.get('blobs', [])
            
            print(f'📊 Found {len(blobs)} blobs in storage')
            return blobs
            
        except requests.RequestException as e:
            print(f'❌ Error fetching blob list: {e}')
            raise
    
    def delete_blob(self, url: str) -> bool:
        """Delete a single blob"""
        try:
            headers = {
                'Authorization': f'Bearer {self.token}',
                'Content-Type': 'application/json'
            }
            
            # Extract blob key from URL
            blob_key = url.split('/')[-1]
            
            response = requests.delete(
                f'https://api.vercel.com/v2/blob/{blob_key}',
                headers=headers
            )
            response.raise_for_status()
            return True
            
        except requests.RequestException as e:
            print(f'❌ Failed to delete {url}: {e}')
            return False
    
    def delete_blobs_batch(self, urls: List[str]) -> Dict[str, Any]:
        """Delete a batch of blobs"""
        results = {
            'successful': 0,
            'failed': 0,
            'errors': []
        }
        
        action = '[DRY RUN] Would delete' if self.dry_run else 'Deleting'
        print(f'🗑️  {action} batch of {len(urls)} blobs...')
        
        for url in urls:
            if self.dry_run:
                print(f'✅ [DRY RUN] Would delete: {url}')
                results['successful'] += 1
            else:
                if self.delete_blob(url):
                    print(f'✅ Deleted: {url}')
                    results['successful'] += 1
                else:
                    results['failed'] += 1
                    results['errors'].append(url)
        
        return results
    
    def format_bytes(self, bytes_size: int) -> str:
        """Format bytes into human readable format"""
        if bytes_size == 0:
            return '0 B'
        
        sizes = ['B', 'KB', 'MB', 'GB']
        k = 1024
        i = 0
        
        while bytes_size >= k and i < len(sizes) - 1:
            bytes_size /= k
            i += 1
        
        return f'{bytes_size:.2f} {sizes[i]}'
    
    def delete_all_blobs(self):
        """Main function to delete all blobs"""
        try:
            # List all blobs
            blobs = self.list_all_blobs()
            
            if not blobs:
                print('✅ No blobs found to delete')
                return
            
            # Show blob information
            print('\n📝 Blob details:')
            total_size = 0
            
            for i, blob in enumerate(blobs, 1):
                size = blob.get('size', 0)
                total_size += size
                uploaded_at = blob.get('uploadedAt', 'Unknown')
                pathname = blob.get('pathname', blob.get('url', 'Unknown'))
                
                print(f'  {i}. {pathname} ({self.format_bytes(size)}, uploaded: {uploaded_at})')
            
            print(f'\n📊 Total size: {self.format_bytes(total_size)}')
            
            if self.dry_run:
                print('\n🔍 DRY RUN MODE: No blobs will be actually deleted')
            else:
                # Confirm deletion
                if not self.confirm_deletion():
                    print('❌ Deletion cancelled by user')
                    return
            
            # Delete blobs in batches
            urls = [blob.get('url', '') for blob in blobs if blob.get('url')]
            total_results = {
                'successful': 0,
                'failed': 0,
                'errors': []
            }
            
            for i in range(0, len(urls), self.batch_size):
                batch = urls[i:i + self.batch_size]
                batch_number = i // self.batch_size + 1
                total_batches = (len(urls) + self.batch_size - 1) // self.batch_size
                
                print(f'\n🔄 Processing batch {batch_number}/{total_batches}')
                
                results = self.delete_blobs_batch(batch)
                
                total_results['successful'] += results['successful']
                total_results['failed'] += results['failed']
                total_results['errors'].extend(results['errors'])
                
                # Delay between batches
                if i + self.batch_size < len(urls):
                    print(f'⏳ Waiting {self.delay_between_batches}s before next batch...')
                    time.sleep(self.delay_between_batches)
            
            # Summary
            print('\n📊 DELETION SUMMARY:')
            action = 'would delete' if self.dry_run else 'deleted'
            print(f'✅ Successfully {action}: {total_results["successful"]} blobs')
            print(f'❌ Failed: {total_results["failed"]} blobs')
            
            if total_results['errors']:
                print('\n❌ Errors:')
                for url in total_results['errors']:
                    print(f'  - {url}')
            
            if not self.dry_run and total_results['successful'] > 0:
                print(f'\n✅ Successfully deleted {total_results["successful"]} blobs from Vercel blob storage')
                print('💡 The blob store should now be empty and can be deleted from the Vercel dashboard')
        
        except Exception as e:
            print(f'💥 Fatal error: {e}')
            sys.exit(1)
    
    def run(self):
        """Run the cleanup tool"""
        print('🚀 Vercel Blob Storage Cleanup Tool (Python)')
        print('============================================\n')
        
        if self.dry_run:
            print('🔍 Running in DRY RUN mode - no actual deletions will occur\n')
        
        self.delete_all_blobs()

def main():
    parser = argparse.ArgumentParser(description='Delete all blobs from Vercel blob storage')
    parser.add_argument('--dry-run', action='store_true', help='Show what would be deleted without deleting')
    parser.add_argument('--confirm', action='store_true', help='Skip confirmation prompt')
    
    args = parser.parse_args()
    
    cleaner = PythonBlobCleaner(
        dry_run=args.dry_run,
        skip_confirmation=args.confirm
    )
    cleaner.run()

if __name__ == '__main__':
    main()

File: scripts/delete-all-blobs.js
#!/usr/bin/env node

/**
 * Delete All Blobs Script for Vercel Blob Storage
 * 
 * This script will delete ALL blobs from your Vercel blob store.
 * Use with extreme caution as this action is irreversible!
 * 
 * Usage:
 *   node scripts/delete-all-blobs.js [--dry-run] [--confirm]
 * 
 * Options:
 *   --dry-run    Show what would be deleted without actually deleting
 *   --confirm    Skip confirmation prompt (use for automation)
 */

const { list, del } = require('@vercel/blob')
const readline = require('readline')

// Configuration
const BATCH_SIZE = 50 // Number of blobs to delete in each batch
const DELAY_BETWEEN_BATCHES = 1000 // 1 second delay between batches

class BlobCleaner {
  constructor() {
    this.token = process.env.BLOB_READ_WRITE_TOKEN
    this.dryRun = process.argv.includes('--dry-run')
    this.skipConfirmation = process.argv.includes('--confirm')
    
    if (!this.token) {
      console.error('❌ Error: BLOB_READ_WRITE_TOKEN environment variable is required')
      process.exit(1)
    }
  }

  async confirmDeletion() {
    if (this.skipConfirmation) {
      return true
    }

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    })

    return new Promise((resolve) => {
      rl.question(
        '⚠️  This will DELETE ALL BLOBS from your Vercel blob store. This action is IRREVERSIBLE!\n' +
        'Are you absolutely sure you want to continue? Type "DELETE ALL" to confirm: ',
        (answer) => {
          rl.close()
          resolve(answer === 'DELETE ALL')
        }
      )
    })
  }

  async listAllBlobs() {
    console.log('📋 Fetching list of all blobs...')
    
    try {
      const { blobs } = await list({
        token: this.token,
        limit: 1000 // Get up to 1000 blobs at once
      })

      console.log(`📊 Found ${blobs.length} blobs in storage`)
      return blobs
    } catch (error) {
      console.error('❌ Error fetching blob list:', error.message)
      throw error
    }
  }

  async deleteBlobsBatch(urls) {
    const results = {
      successful: 0,
      failed: 0,
      errors: []
    }

    console.log(`🗑️  Deleting batch of ${urls.length} blobs...`)

    for (const url of urls) {
      try {
        if (!this.dryRun) {
          await del(url, { token: this.token })
        }
        results.successful++
        console.log(`✅ ${this.dryRun ? '[DRY RUN] Would delete' : 'Deleted'}: ${url}`)
      } catch (error) {
        results.failed++
        results.errors.push({ url, error: error.message })
        console.error(`❌ Failed to delete ${url}: ${error.message}`)
      }
    }

    return results
  }

  async deleteAllBlobs() {
    try {
      // List all blobs
      const blobs = await this.listAllBlobs()
      
      if (blobs.length === 0) {
        console.log('✅ No blobs found to delete')
        return
      }

      // Show blob information
      console.log('\n📝 Blob details:')
      blobs.forEach((blob, index) => {
        const sizeKB = (blob.size / 1024).toFixed(2)
        console.log(`  ${index + 1}. ${blob.pathname} (${sizeKB} KB, uploaded: ${new Date(blob.uploadedAt).toLocaleString()})`)
      })

      // Calculate total size
      const totalSize = blobs.reduce((sum, blob) => sum + blob.size, 0)
      const totalSizeMB = (totalSize / 1024 / 1024).toFixed(2)
      console.log(`\n📊 Total size: ${totalSizeMB} MB`)

      if (this.dryRun) {
        console.log('\n🔍 DRY RUN MODE: No blobs will be actually deleted')
      } else {
        // Confirm deletion
        const confirmed = await this.confirmDeletion()
        if (!confirmed) {
          console.log('❌ Deletion cancelled by user')
          return
        }
      }

      // Delete blobs in batches
      const urls = blobs.map(blob => blob.url)
      const totalResults = {
        successful: 0,
        failed: 0,
        errors: []
      }

      for (let i = 0; i < urls.length; i += BATCH_SIZE) {
        const batch = urls.slice(i, i + BATCH_SIZE)
        const batchNumber = Math.floor(i / BATCH_SIZE) + 1
        const totalBatches = Math.ceil(urls.length / BATCH_SIZE)
        
        console.log(`\n🔄 Processing batch ${batchNumber}/${totalBatches}`)
        
        const results = await this.deleteBlobsBatch(batch)
        
        totalResults.successful += results.successful
        totalResults.failed += results.failed
        totalResults.errors.push(...results.errors)

        // Delay between batches to avoid rate limiting
        if (i + BATCH_SIZE < urls.length) {
          console.log(`⏳ Waiting ${DELAY_BETWEEN_BATCHES}ms before next batch...`)
          await new Promise(resolve => setTimeout(resolve, DELAY_BETWEEN_BATCHES))
        }
      }

      // Summary
      console.log('\n📊 DELETION SUMMARY:')
      console.log(`✅ Successfully ${this.dryRun ? 'would delete' : 'deleted'}: ${totalResults.successful} blobs`)
      console.log(`❌ Failed: ${totalResults.failed} blobs`)
      
      if (totalResults.errors.length > 0) {
        console.log('\n❌ Errors:')
        totalResults.errors.forEach(({ url, error }) => {
          console.log(`  - ${url}: ${error}`)
        })
      }

      if (!this.dryRun && totalResults.successful > 0) {
        console.log(`\n✅ Successfully deleted ${totalResults.successful} blobs from Vercel blob storage`)
        console.log('💡 The blob store should now be empty and can be deleted from the Vercel dashboard')
      }

    } catch (error) {
      console.error('💥 Fatal error:', error.message)
      process.exit(1)
    }
  }

  async run() {
    console.log('🚀 Vercel Blob Storage Cleanup Tool')
    console.log('=====================================\n')

    if (this.dryRun) {
      console.log('🔍 Running in DRY RUN mode - no actual deletions will occur\n')
    }

    await this.deleteAllBlobs()
  }
}

// Run the script
if (require.main === module) {
  const cleaner = new BlobCleaner()
  cleaner.run().catch(console.error)
}

module.exports = BlobCleaner
